### DISCUSSION TITLE
Need suggestions for my approach of putting cef related files into subfolder

### DISCUSSION THREAD

**Discussion by RadarNyan** [cefsharp/CefSharp#3371] [Show and tell]
By default cef puts all the files in the root folder of the program, which is kinda messy.
Over the years there has been some mentioning of it like #601 but those information are fairly old.

For now here's my practice after gathering many informations from old issues:

1. Create a .Net 4.5.2 WinForms project (```cef_in_subfolder```) from Visual Studio 2019.
2. Add CefSharp.WinForms from nuget (at the time of writing of this, 86.0.241 stable).
3. Change target platform to x86, save everything and close the solution.

Then I have to manually edit two files:
1. Edit ```cef_in_subfolder.csproj``` and add following code to the first ```<PropertyGroup>```
```xml
    <CefSharpTargetDir>cef\bin</CefSharpTargetDir>
``` 
2. Edit ```App.config``` and add following code to the ```<configuration>``` block
```xml
    <runtime> 
        <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
            <probing privatePath="cef\bin"/>
        </assemblyBinding>
    </runtime>
```

Basically at this point it's done, here's my ```Form1.cs```
```csharp
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.IO;
using CefSharp;
using CefSharp.WinForms;

namespace cef_in_subfolder
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
            
            // load cef
            var cef_path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "cef");
            var cef_settings = new CefSettings()
            {
                LogFile = Path.Combine(cef_path, "cef_log.txt"),
                CachePath = Path.Combine(cef_path, "cache"),
                UserDataPath = Path.Combine(cef_path, "userdata")
            };
            Cef.Initialize(cef_settings, performDependencyCheck: false, browserProcessHandler: null);

            Controls.Add(new ChromiumWebBrowser(@"https://www.google.com/"));
        }
    }
}

```

 I find that it's unnecessary to specify ```BrowserSubprocessPath```, ```LocalesDirPath``` or ```ResourcesDirPath``` as cef works just fine and I've tested changing Locale works too.

---

It works and gives a very clean program root folder. However I didn't fully understand how CefSharp works thus I need some suggestions about this approach, mainly if there is something that's broken just I haven't tested it yet, and I'd love to know if there is a better approach.

**Comment by amaitland** [comment]
> I find that it's unnecessary to specify BrowserSubprocessPath, LocalesDirPath or ResourcesDirPath as cef works just fine and I've tested changing Locale works too.

The browsersubprocess is now defaulted to the same location as the CefSharp.Core.dll, this in most instances is correct and only needs changing in rare cases.

> It works and gives a very clean program root folder. However I didn't fully understand how CefSharp works thus I need some suggestions about this approach

Looks fine. Did you have a specific question? 

Be aware that that version 87 will bring some changes to the nuget packages see #3319 for details.

**Reply by RadarNyan** [reply]
> Be aware that that version 87 will bring some changes to the nuget packages see #3319 for details.

I tried to repeat what I wrote above with version 87, found that specifying ```<CefSharpTargetDir>cef\bin</CefSharpTargetDir>``` still copies some but not all cef related files. This is mentioned in #3319 , in my case```CefSharp.dll```, ```CefSharp.Core.dll``` and ```CefSharp.Winforms.dll``` are left at root folder and need to be manually moved which I did with Post-build event:

```cmd
move /Y "$(TargetDir)CefSharp.*" "$(TargetDir)cef\bin\"
```

However ```CefSHarp.Core.Runtime.dll``` has a duplication in root folder as well as the folder I set as ```CefSharpTargetDir```, I've compared checksum to make sure they are the same file. Is this some kind of an overlook?

**Reply by amaitland** [reply]
> However CefSHarp.Core.Runtime.dll has a duplication in root folder as well as the folder I set as CefSharpTargetDir, I've compared checksum to make sure they are the same file. Is this some kind of an overlook?

An extra check will need to be added to https://github.com/cefsharp/CefSharp/blob/master/NuGet/CefSharp.Common.targets#L4 for the case where CefSharpTargetDir is not empty string. 

Pull requests welcome.

**Reply by amaitland** [reply]
This should be fixed in https://github.com/cefsharp/CefSharp/commit/2b2031ef036b6181c3130d63dc57e9fb17b19371#diff-4d4db7f7d2784381f23dcebde8e2769047ae8465335575e9872cf666e8b9900aR9

**Comment by BuslikDrev** [comment]
У меня CefSharp 109 в Windwos 7 работает probing privatePath, а в Windows 10 уже пишет:
```
Необработанное исключение: System.Reflection.TergetInvocationException: Адресат вызова создал исключение. ---> System.IO.FileNotFoundException: Не удалось загрузить файл или сборку "CefSharp.Core.Runtime.dll" либо одну из их зависимостей. Не найден указанный модуль.
```

```
AssemblyLoad... CefSharp.WinForms, Version=109.1.110.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138
AssemblyLoad... CefSharp, Version=109.1.110.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138
AssemblyLoad... CefSharp.Core, Version=109.1.110.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138
FirstChanceException... System.IO.FileNotFoundException: Не удалось загрузить файл или сборку "CefSharp.Core.Runtime.dll" либо одну из их зависимостей. Не найден указанный модуль.
Имя файла: 'CefSharp.Core.Runtime.dll'
   в CefSharp.CefSettingsBase..ctor()
```

Помогла установка VC++ x86 и x64:
https://learn.microsoft.com/ru-ru/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022
